### DISCUSSION TITLE
Z offset constantly changing

### DISCUSSION THREAD

**Discussion by Snipermechanic** [Jyers/Marlin#1782] [General Discussion]
Running Ender 3 V2, with latest FW.

Issue is that I go through and do all levelling etc, set offset and do a print. Perfection.

Next print, my offset shows as correct (-0.16) but nozzle it to far away and I have to do -0.32, this goes on and on.

no matter what I do, the offset is forever changing. Even turning it off between prints.

It's not a mechanical issue, i've rebuilt and checked printer 10x times now.

Any thoughts?

**Comment by K2Kevin** [comment]
You are using a BLtouch or CR-Touch?

**Reply by Snipermechanic** [reply]
Not using a sensor. Just gone for 3x3 mesh levelling

**Reply by K2Kevin** [reply]
I never dealt with Z-Offset until I added a ABL (Cr-Touch).  
You should set it to ZERO. and adjust your z-limit switch ONCE to match your bed before leveling (If its low). Should be a bit higher than your current bed to allow for manual leveling. After you manually level, then run your manual mesh level, that tweaks the Z-offset for you. At least that's what worked flawlessly for me.  Never set a z-offset until I added the CR-Touch.

**Reply by Snipermechanic** [reply]
The Offset is needed as the mesh is fine for levelling, but it doesn't allow you to tune first layer, which is what the offset/live Z adjust is for.

**Reply by K2Kevin** [reply]
Not what I remember manual mesh leveling does.  Thats the only place you make manual changes to the z-offet at all your mesh positions.  NOT the main z-offset setting in the menus. The MAIN z-offset was always set to zero and I never touched it.

**Reply by Snipermechanic** [reply]
Doing the manual mesh is fine as it tunes for inconsistency in the bed. Live Z in other printers i've owned was specifically to allow you to tune your first layer, or sdet offsets for different filaments/beds/sheets etc.

TBH I don't even think the mesh levelling works

**Reply by K2Kevin** [reply]
1. Set our z-offset to ZERO and save.
2. Do Manual Bed Leveling (adjusting the knobs under the bed)
3. Goto level menu and do your manual mesh (Create New Mesh).
4. Now do your fine adjustments with paper and do your micro step up/down. SAVE MESH.

Manual mesh works and I've used it. 
 **I wonder if you haven't changed the G-Code in Cura so it actually loads your mesh. **M420 S1**** 

I know there is a good video out there of this that might help you.  Ill look for it.

Found one:  https://www.youtube.com/watch?v=mlnA73vgoUk&t=509s

**Reply by Snipermechanic** [reply]
I've done this, many many times.

Just setting the mesh, isn't enough. I fully understand it's purpose. I've got it enabled in my Gcode too.

As per this video: https://www.youtube.com/watch?v=mlnA73vgoUk&t=567s

I've done all of this, however the setting appears to be lost after each print.

I'm trying to work out if it's a known bug, if there's a specific Gcode or something else

**Reply by K2Kevin** [reply]
> Doing the manual mesh is fine as it tunes for inconsistency in the bed. Live Z in other printers i've owned was specifically to allow you to tune your first layer, or sdet offsets for different filaments/beds/sheets etc.

So is that what you are doing, changing the bed sheets each time you boot machine and having to manually change the Z-offset?  Didn't think so.

**Reply by K2Kevin** [reply]
Show me your Start G-code...

FYI, you know you have to re-slice all your old saved prints with the new G-code in order to use the Mesh.

**Reply by Snipermechanic** [reply]
I have done exactly as you have said. I'm not disputing your advice and I absolutely appreciate the help. I'm simply trying to explain my issue, outside of doing what you've advised and I've already done.

The first layer is not squishing into the textured sheet as much as it needs to (hence the live Z/offset).

The nozzle is ever so slightly to far away from the bed, regardless of how many times it's levelled, with paper or a feeler gauge.

So while the bed is level and the mesh set up to account for any inconsistencies in the bed. The issue still remains, to get the first layer squish I need, I need to live adjust my Z (one of the features of the firmware).

While doing my layer test, I adjust the z live (to say -0.16), I get the perfect first layer, great adhesion, very smooth.

The I print again straight after, or after turning machine off and, the offset still shows, but it's no longer applied, it shows me being at -0.16, but the nozzle is to far away from bed etc.

frustrating! lol

**Comment by Snipermechanic** [comment]
So a full breakdown.

I do full level/setup,. offset etc.

I print and if it's the first print since machine has been turned of, it will apply the correct offset.

Next print however, the offset will show on the LCD, but i'll have to use live Z to adjust.

Printer heats up, does the purge line and I have a split second just as the skirt starts, to quickly dial in the Z. Whether I save at any stage seems to make 0 difference, the next print will need the offset again.

I can literally start at -0.16 and get perfect print, then next print will be -0.32, then -0.48 and so on. I save settings from the perfect -0.16.

Printer goes on, I do next print, LCD shows the -0.167, but it's not applied

**Reply by Snipermechanic** [reply]
I don't know how :(

**Reply by K2Kevin** [reply]
This only effects those without a BLtouch ?

**Reply by Snipermechanic** [reply]
> The fix : in Marlin/src/lcd/e3v2/jyersui/dwin.cpp :
> 
> ```
> #if HAS_ZOFFSET_ITEM
>     static float lastzoffset = zoffsetvalue;
>    Modify -> if (zoffsetvalue != lastzoffset) {
>   By ->  if (zoffsetvalue != lastzoffset && !printing) {
>       lastzoffset = zoffsetvalue;
>       #if HAS_BED_PROBE
>         probe.offset.z = zoffsetvalue;
> ```

I've seen a guy called Bryan Vines who's got a video on YT about compiling. I will try and learn it tomorrow. If I get stuck and I OK to pick your brain? 

The quality and every other aspect has been fine, but this resetting offset has been driving me mad. I went with the 4.2.2 with manual 3x3 and its been fine minus this issue.

**Reply by Snipermechanic** [reply]
> 

I've tried to compile following this video: https://www.youtube.com/watch?v=so0FfFVEmiw&t=315s

I've found the #if HAS_ZOFFSET_ITEM line and updated it to match what you've put. However the compile fails with:

Environment                   Status    Duration
----------------------------  --------  ------------
STM32F103RET6_creality_maple  FAILED    00:00:06.129
=======================================================
